### PR TITLE
planner: move flagPrunColumnsAgain to the last

### PR DESCRIPTION
**Pull request by SunRunAway**
### What problem does this PR solve? 

part of  #14428
A follower PR of #14683

Some rules (like `joinReOrderSolver`) may rebuild the schema of a logical plan, and it will overwrite the previous efforts of column pruning. Thus, `flagPrunColumnsAgain` should be the last optimization rule.

### What is changed and how it works?

1. move `flagPrunColumnsAgain` to the last
2. implement `LogicalTopN.PruneColumns` since `flagPrunColumnsAgain` is after `flagPushDownTopN` now.
3. Fix the implementation of `LogicalUnionAll.PruneColumns`, to make sure keeping the schema of `LogicalUnionAll` same as its children's
4. Fix the implementation of `partitionProcessor`, to make sure the schema of `UnionAll` is not referenced by its children's

### Check List 

Tests 

 - Unit test

Code changes

 - None

Side effects

 - None

### CODE REVIEW

File: cmd/explaintest/r/tpch.result [hunk 1]
```diff
@@ -926,9 +926,9 @@ p_brand,
 p_type,
 p_size;
 id	count	task	operator info
-Sort_13	3863988.24	root	Column#23:desc, tpch.part.p_brand:asc, tpch.part.p_type:asc, tpch.part.p_size:asc
-└─Projection_15	3863988.24	root	tpch.part.p_brand, tpch.part.p_type, tpch.part.p_size, Column#23
-  └─HashAgg_18	3863988.24	root	group by:tpch.part.p_brand, tpch.part.p_size, tpch.part.p_type, funcs:count(distinct tpch.partsupp.ps_suppkey)->Column#23, funcs:firstrow(tpch.part.p_brand)->tpch.part.p_brand, funcs:firstrow(tpch.part.p_type)->tpch.part.p_type, funcs:firstrow(tpch.part.p_size)->tpch.part.p_size
+Sort_13	14.41	root	Column#23:desc, tpch.part.p_brand:asc, tpch.part.p_type:asc, tpch.part.p_size:asc
+└─Projection_15	14.41	root	tpch.part.p_brand, tpch.part.p_type, tpch.part.p_size, Column#23
+  └─HashAgg_18	14.41	root	group by:tpch.part.p_brand, tpch.part.p_size, tpch.part.p_type, funcs:count(distinct tpch.partsupp.ps_suppkey)->Column#23, funcs:firstrow(tpch.part.p_brand)->tpch.part.p_brand, funcs:firstrow(tpch.part.p_type)->tpch.part.p_type, funcs:firstrow(tpch.part.p_size)->tpch.part.p_size
```
Comment (francis0407) [inline cmd/explaintest/r/tpch.result]: Do you know why is this changed?

File: planner/core/rule_column_pruning.go [hunk 2]
```diff
@@ -151,10 +151,30 @@ func (ls *LogicalSort) PruneColumns(parentUsedCols []*expression.Column) error {
 	return child.PruneColumns(parentUsedCols)
 }
 
+// PruneColumns implements LogicalTopN interface.
```
Comment (francis0407) [inline planner/core/rule_column_pruning.go]:
```suggestion
// PruneColumns implements LogicalPlan interface.
```

File: cmd/explaintest/r/tpch.result (hunk 1 above)
Comment (SunRunAway) [reply cmd/explaintest/r/tpch.result]: I'm investigating.

File: planner/core/rule_column_pruning.go [hunk 3]
```diff
@@ -151,10 +151,30 @@ func (ls *LogicalSort) PruneColumns(parentUsedCols []*expression.Column) error {
...
+		cols := expression.ExtractColumns(lt.ByItems[i].Expr)
+		if len(cols) == 0 {
+			if expression.IsMutableEffectsExpr(lt.ByItems[i].Expr) {
+				continue
+			}
+			lt.ByItems = append(lt.ByItems[:i], lt.ByItems[i+1:]...)
+		} else if lt.ByItems[i].Expr.GetType().Tp == mysql.TypeNull {
+			lt.ByItems = append(lt.ByItems[:i], lt.ByItems[i+1:]...)
+		} else {
+			parentUsedCols = append(parentUsedCols, cols...)
```
Comment (lzmhhh123) [inline planner/core/rule_column_pruning.go]: Should we check whether the columns are in `parentUsedCols`?

Comment (SunRunAway) [reply planner/core/rule_column_pruning.go]: Both can. The redundant columns will not affect correctness.
